### PR TITLE
[prebuild-config] add support for iOS 18 dark/tinted icons

### PR DESCRIPTION
**Pull request by fobos531**
# Why

iOS 18 introduced support for dark/tinted icons in the launcher/home screen. This PR introduces an opt-in, backwards-compatible pattern for users to specify a dark/tinted variant of their icon in case they happen to build their apps for iOS 18. Building on current EAS infrastructure (Xcode 15) is not affected, it still works fine. The implementation has been tested on both Xcode 15 and 16

# How

Made the necessary adjustments to the prebuild logic.

# Test Plan

Updated the test suites. I haven't actually tested the prebuild process, since I'm not familiar how to test this (I've only been contributing to the native SDK packages, so any guidance here would be appreciated).

Also cc @amandeepmittal @Simek - how can I regenerate the `app-config-schema.json` so this is accurately reflected in the docs?

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by amandeepmittal** [comment]
> Also cc @amandeepmittal @Simek - how can I regenerate the `app-config-schema.json` so this is accurately reflected in the docs?

Hi @fobos531, we currently keep local copies of the app config schema (based on versions) in a private repo because it syncs with the modules API. Can you please provide me details or point me towards it? I'll be happy to follow up in a separate PR.

**Comment by fobos531** [comment]
> > Also cc @amandeepmittal @Simek - how can I regenerate the `app-config-schema.json` so this is accurately reflected in the docs?
> 
> Hi @fobos531, we currently keep local copies of the app config schema (based on versions) in a private repo because it syncs with the modules API. Can you please provide me details or point me towards it? I'll be happy to follow up in a separate PR.

Hello @amandeepmittal , could you clarify about what exactly your inquiry is? If you're asking as to what exactly would need to be changed in the schema, the `ios.icon` field can now optionally be an object of three strings (any, dark, tinted) which denote which icon should be used for a given iOS appearance setting.

**Comment by amandeepmittal** [comment]
> > > Also cc @amandeepmittal @Simek - how can I regenerate the `app-config-schema.json` so this is accurately reflected in the docs?
> > 
> > 
> > Hi @fobos531, we currently keep local copies of the app config schema (based on versions) in a private repo because it syncs with the modules API. Can you please provide me details or point me towards it? I'll be happy to follow up in a separate PR.
> 
> Hello @amandeepmittal , could you clarify about what exactly your inquiry is? If you're asking as to what exactly would need to be changed in the schema, the `ios.icon` field can now optionally be an object of three strings (any, dark, tinted) which denote which icon should be used for a given iOS appearance setting.

Yeah, that's what I was looking for. Thanks for the information!

**Comment by fobos531** [comment]
Hello @douglowder @EvanBacon, do you perhaps have the bandwith to check this out?

File: packages/@expo/config-types/build/ExpoConfig.d.ts
Comment (douglowder) [inline packages/@expo/config-types/build/ExpoConfig.d.ts:297]: Unfortunately this is not the correct way to change the types in the Expo config:
- The types are actually derived from the internal schema used in our services, so the next schema update would wipe out this change.
- If you did change them here manually, they would have to be changed in the source (`packages/@expo/config-types/src/ExpoConfig.ts`) and then the module must be rebuilt.

**Review by douglowder — CHANGES_REQUESTED** [review]
We cannot merge this, for the reasons described in the inline comments.

We will discuss this internally and see when we would be able to add support for this feature.

File: packages/@expo/config-types/build/ExpoConfig.d.ts
Comment (fobos531) [reply packages/@expo/config-types/build/ExpoConfig.d.ts:297]: I understand, no worries. I figured that is the case based on @amandeepmittal 's comments. In the very least, I hope this PR helps steer the Expo team in a good direction regarding implementing support for this. Thanks for taking a look at this nevertheless! Feel free to close the PR if needed.

Comment (douglowder) [reply packages/@expo/config-types/build/ExpoConfig.d.ts:297]: We appreciate the heads up about this feature, and thanks for the PR! 👍

**Comment by iLynxcat** [comment]
@douglowder Any opposition to me opening an issue (or discussion) to officially keep track of this feature? I am eyeballing this for @spacedriveapp and would love somewhere o keep subscribed to for this in particular

**Comment by douglowder** [comment]
> @douglowder Any opposition to me opening an issue (or discussion) to officially keep track of this feature? I am eyeballing this for @spacedriveapp and would love somewhere to keep subscribed to for this in particular

Absolutely, feel free to create an issue (please use the "Feature Request" category) and link to it here, and I will link that to our internal work item for this.

**Comment by iLynxcat** [comment]
> > @douglowder Any opposition to me opening an issue (or discussion) to officially keep track of this feature? I am eyeballing this for @spacedriveapp and would love somewhere to keep subscribed to for this in particular
> 
> Absolutely, feel free to create an issue (please use the "Feature Request" category) and link to it here, and I will link that to our internal work item for this.

Created! Added a demo image to help explain the expected icon set input & resulting icons as displayed in iOS. https://expo.canny.io/feature-requests/p/add-support-for-dark-tinted-app-icons-on-ios

**Comment by brentvatne** [comment]
hey @fobos531, thank you for the pr! we'll take a look at this next week. we can easily copy over any changes to our internal copy of the schema :)

**Comment by Kaique292** [comment]
Any update about this?

**Comment by CodeWithShreyans** [comment]
iOS 18 is coming out in 2 weeks, any updates on this?

**Comment by vitorleo80** [comment]
Any updates on this issue? Customers are already requesting this in our apps. Thanks!

**Comment by fobos531** [comment]
cc @brentvatne @alanjhughes

**Comment by alanjhughes** [comment]
@fobos531 - This looks ok but needs testing before merging. You can use the `sandbox` app to generate a new project and test the results. We'll also need to test xcode 15 compatibility. You can download multiple versions of xcode so should be easy to verify

**Comment by alanjhughes** [comment]
Also note that this feature is still pretty buggy on the latest betas and we typically don't add support for the latest OS until it is stable

**Comment by CodeWithShreyans** [comment]
@alanjhughes that's reasonable but a lack of this feature could result in users seeing "not correct" icons for apps when iOS 18 comes out

**Comment by alanjhughes** [comment]
@CodeWithShreyans it could but, there are many apps that won't support this initially. This is going to be a feature that has slow adoption because it's not critical to have it. There's also the question of if users will even use it. In any case, it's not going to be a deal breaker for apps if they don't support this on day one. I don't think even apple expect wide spread day one adoption which is why they provide a default for the apps that are not ready.

**Comment by fobos531** [comment]
> @alanjhughes that's reasonable but a lack of this feature could result in users seeing "not correct" icons for apps when iOS 18 comes out

I think @alanjhughes raises some valid concerns. 

Apple already has a "best effort" system in place where it tries to theme your icon out of the box if they do not support tinting/dark mode, and I found that it works pretty well. Of course, it doesn't work well on all applications, so thats why it is necessary to support this feature.

> Also note that this feature is still pretty buggy on the latest betas

That might be the case, but the important thing is that the actual API for defining such icons, nor the guidelines have changed since start of summer, basically. In that sense this API could be considered "stable" and 99% will not change at all in the remaining time until the release. 

> it could but, there are many apps that won't support this initially. This is going to be a feature that has slow adoption because it's not critical to have it. There's also the question of if users will even use it. In any case, it's not going to be a deal breaker for apps if they don't support this on day one. I don't think even apple expect wide spread day one adoption which is why they provide a default for the apps that are not ready.

I agree with this. Its basically the same thing like there is with the `monochromeIcon` on Android. It wasn't adopted by Expo straight away, but a little later, as its mainly a mechanism to provide visual consistency on your home screen (it looks much better if all icons follow material you theming instead of having non-complying icons)

In any case, I will follow the developments regarding this, and try to keep this PR up-to-date and test it out when I find available time, and report back

**Comment by fobos531** [comment]
@alanjhughes 

Testing results: 

No icon config set

```
"ios": {
      "supportsTablet": true,
      "bundleIdentifier": "com.jakov.glavina.sandbox"
    },
```

Xcode 15 & iOS 17.5

Light & Dark mode: 

CleanShot 2024-08-30 at 17 00 29@2x

Xcode 16 & iOS 18.0:

"Automatic" icon appearance:

Light & Dark mode:

CleanShot 2024-08-30 at 17 03 33@2x

For "Light" icon appearance it is the same as "Automatic".

"Dark" appearance just tints an image both in Light & Dark mode: 

CleanShot 2024-08-30 at 17 05 42@2x

"Tinted" icon appearance tints an image according to the selected color in both Light & Dark mode:

CleanShot 2024-08-30 at 17 07 00@2x

"Any" icon specified: 

```
"ios": {
      "supportsTablet": true,
      "bundleIdentifier": "com.jakov.glavina.sandbox",
      "icon": {
        "any": "./launcher.png"
      }
    },
```

Used icon: 

launcher

Xcode 15 & iOS 17.5

Light & Dark mode: 

CleanShot 2024-08-30 at 17 12 59@2x

Xcode 16 & iOS 18.0:

"Automatic" icon appearance:

Light mode:

CleanShot 2024-08-30 at 17 17 05@2x

Dark mode: 

CleanShot 2024-08-30 at 17 17 05@2x

"Light" icon appearance:

Light mode:

CleanShot 2024-08-30 at 17 18 04@2x

Dark mode:

CleanShot 2024-08-30 at 17 18 25@2x

"Dark" icon appearance:

Light mode:

CleanShot 2024-08-30 at 17 18 50@2x

Dark mode:

CleanShot 2024-08-30 at 17 18 50@2x

"Tinted" icon appearance:

Light & Dark mode:

CleanShot 2024-08-30 at 17 21 53@2x


Any/Dark/Tinted config: 

```
   "ios": {
      "supportsTablet": true,
      "bundleIdentifier": "com.jakov.glavina.sandbox",
      "icon": {
        "any": "./light.png",
        "dark": "./dark.png",
        "tinted": "./tinted.png"
      }
    },
```

Icons used:

Any:
light


Dark:

dark

Tinted:

tinted

Xcode 15 & iOS 17.5

Light & Dark mode: 

CleanShot 2024-08-30 at 17 47 11@2x

Xcode 16 & iOS 18.0:

"Automatic" icon appearance:

Light mode:

CleanShot 2024-08-30 at 17 50 07@2x

Dark mode:

CleanShot 2024-08-30 at 17 50 13@2x


"Light" icon appearance:

Light mode:

CleanShot 2024-08-30 at 17 50 13@2x

Dark mode: 

CleanShot 2024-08-30 at 17 50 44@2x

"Dark" icon appearance **(please note that I didn't have a better "dark" icon at disposal at the time of testing, but this can still be used to showcase that specifically this icon is being used with the "dark" icon appearance, meaning it works fine)**:

Light mode & Dark mode:

CleanShot 2024-08-30 at 17 51 24@2x


"Tinted" icon appearance:

Light & Dark mode:

CleanShot 2024-08-30 at 17 51 24@2x


Example of generated icons structure:

CleanShot 2024-08-30 at 17 54 41@2x

Contents.json (this works fine whether you run it on Xcode 15 or 16

```
{
  "images": [
    {
      "filename": "App-Icon-1024x1024@1x.png",
      "idiom": "universal",
      "platform": "ios",
      "size": "1024x1024"
    },
    {
      "filename": "App-Icon-dark-1024x1024@1x.png",
      "idiom": "universal",
      "platform": "ios",
      "size": "1024x1024",
      "appearances": [
        {
          "appearance": "luminosity",
          "value": "dark"
        }
      ]
    },
    {
      "filename": "App-Icon-tinted-1024x1024@1x.png",
      "idiom": "universal",
      "platform": "ios",
      "size": "1024x1024",
      "appearances": [
        {
          "appearance": "luminosity",
          "value": "tinted"
        }
      ]
    }
  ],
  "info": {
    "version": 1,
    "author": "expo"
  }
}
```

I highly recommend that you test this internally as well, but from my testing it seems to work fairly well.

Let me know if you need any other input from me.

File: packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts
Comment (zacherya) [inline packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts:61]: These conditions could be combined, no point having so many separate conditions with the same line of executing code.
```suggestion
  if (!icon || (typeof icon === 'string' && !icon) || (typeof icon === 'object' && !icon?.any && !icon?.dark && !icon?.tinted)) {
    WarningAggregator.addWarningIOS('icon', 'No icon is defined in the Expo config.');
  }
```
Hope work gets done in this area soon or this PR gets merged. Thanks for your work :)

**Comment by JustJoostNL** [comment]
Would love to see this in Expo! Even though the adoption at the start may be low, I'm pretty sure that will change quickly!

**Comment by fobos531** [comment]
Hello @alanjhughes - since iOS 18 is officially released, just wondering if there's anything else that would be blocking this PR?

**Comment by alanjhughes** [comment]
Sorry for the delay @fobos531 someone will get back to you on this shortly. Thanks for the great contribution!

File: packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts
Comment (EvanBacon) [inline packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts:93]: tinted icons can have transparent backgrounds but this function removes the transparency from the icon.

https://developer.apple.com/design/human-interface-guidelines/app-icons#Platform-considerations

Comment (fobos531) [reply packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts:93]: @EvanBacon If I'm reading the guidelines correctly, in addition to the tinted icons (where you provide a fully opaque icon), you also provide a transparent icon for the dark appearance.

So it seems like transparency should be removed for the `any` variant, but preserved for `dark` and `tinted` variants. Is this correct?

Comment (CodeWithShreyans) [reply packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts:93]: I think it needs to be removed from `tinted` too. Only preserved in `dark`.

**Comment by fobos531** [comment]
Performed some additional tests - all on iOS 18 and Xcode 16 only (since this is now supported in EAS)

These are the icons I used for testing 

Any:
light

Dark: 
dark


Tinted:
launcher


Automatic appearance:

Light mode:

CleanShot 2024-09-28 at 13 05 53@2x

Dark mode:

CleanShot 2024-09-28 at 13 05 58@2x

Tinted appearance (Light & Dark mode): 

CleanShot 2024-09-28 at 13 07 48@2x

Light appearance (Light & Dark mode):

CleanShot 2024-09-28 at 13 08 05@2x

Dark appearance (Light & Dark mode):

CleanShot 2024-09-28 at 13 08 33@2x

cc @EvanBacon thanks for the heads up about the icon transparency - following some additional comments, I'm pretty confident that actually ONLY "dark" variant needs to have their transparency preserved, while for others transparency can be safely removed as it always was. This can be confirmed by the fact that for both dark & tinted variants the icon seems to be properly themed (previously I thought my actual dark icon was messed up, whereas the issue lied in the transparency stuff in prebuild)


With that said @brentvatne I believe all of the points have been resolved this and at this moment I have nothing else to add from my end. As always, let me know if you need anything else from me.

**Comment by jrhager84** [comment]
Is this being implemented soon? I don't see anything officially. Thanks!

File: packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts
Comment (brentvatne) [inline packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts:123]:
```suggestion
    name =`${name}-${appearance}`;
```

File: packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts
Comment (brentvatne) [inline packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts:126]:
```suggestion
  name = `${name}-${size}x${size}@${scale}x.png`;
```

**Review by brentvatne — APPROVED** [review]
this is fantastic, thank you!

**Review by brentvatne — APPROVED** [review]

**Comment by iLynxcat** [comment]
WOOO thank you!!!!

**Comment by iLynxcat** [comment]
was P3 supported before this change?

**Comment by brentvatne** [comment]
@benkingcode - any interest in opening a PR for that? or an issue that explains the problem?

**Comment by roy-borrowell** [comment]
Great work!! Does anyone know how I can start using this change in my prebuild?

I am using expo": "^50.0.19" and I tried using the new config but couldn't get past prebuilding. Is there a way for me to force prebuilding to use a specific version?

**Comment by brentvatne** [comment]
@roy-borrowell - this will be included in sdk 52 when it is released! if you'd like to use it now, you can try to copy over the changes to a custom config plugin in your project

**Comment by brentvatne** [comment]
@fobos531 - do you mind if i use those logos in the sdk 52 release notes?

**Comment by fobos531** [comment]
> @fobos531 - do you mind if i use those logos in the sdk 52 release notes? to demonstrate your work on tinted icons here

hey @brentvatne, thanks for asking - it might be better to reference the apple's human interface guidelines here instead, they also have examples of such icons

**Comment by brentvatne** [comment]
@fobos531 - no problem, i can come up with another example. just thought it'd be nice to potentially give some subtle promo for your app

**Comment by fobos531** [comment]
> @fobos531 - no problem, i can come up with another example. just thought it'd be nice to potentially give some subtle promo for your app

Hey @brentvatne - all good. That would be great, but since I'm not owner of the said app/logo, I'd rather err on the side of caution. Thanks for understanding :)

File: packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts
Comment (EvanBacon) [reply packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts:93]: Looks like you guys didn't implement this correctly and tinted icons have transparency removed.

**Comment by vishnuc** [comment]
Hi , i use sdk52 , how do i use it ?

**Comment by iLynxcat** [comment]
> Hi , i use sdk52 , how do i use it ?

Please read the documentation https://docs.expo.dev/develop/user-interface/splash-screen-and-app-icon/#custom-configuration-tips-for-android-and-ios

**Pull request merged**
